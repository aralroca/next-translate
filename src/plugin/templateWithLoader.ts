import { overwriteLoadLocales } from './utils'

export default function templateWithLoader(
  rawCode: string,
  {
    page = '',
    typescript = false,
    loader = 'getStaticProps',
    hasLoader = false,
    hasLoadLocaleFrom = false,
  } = {}
) {
  const tokenToReplace = `__CODE_TOKEN_${Date.now().toString(16)}__`
  let modifiedCode = rawCode

  if (hasLoader) {
    modifiedCode = modifiedCode
      // Replacing:
      //    const getStaticProps = () => console.log('getStaticProps')
      //    import getStaticProps from './getStaticProps'
      // To:
      //    const _getStaticProps = () => console.log('getStaticProps')
      //    import _getStaticProps from './getStaticProps'
      .replace(
        new RegExp(
          `(const|var|let|async +function|function|import|import {.* as) +${loader}\\W`
        ),
        (v: string) =>
          v.replace(new RegExp(`\\W${loader}\\W`), (r) =>
            r.replace(loader, '_' + loader)
          )
      )
      // Replacing:
      //    export const _getStaticProps = () => ({ props: {} })
      // To:
      //    const _getStaticProps = () => ({ props: {} })
      .replace(
        new RegExp(
          `export +(const|var|let|async +function|function) +_${loader}`
        ),
        (v: string) => v.replace('export', '')
      )
      // Replacing: "export { getStaticProps }" to ""
      .replace(/export +\{ *(getStaticProps|getServerSideProps)( |,)*\}/, '')
      // Replacing:
      //    export { something, getStaticProps, somethingelse }
      // To:
      //    export { something, somethingelse }
      // And:
      //    export { getStaticPropsFake, somethingelse, b as getStaticProps }
      // To:
      //    export { getStaticPropsFake, somethingelse }
      .replace(
        new RegExp(`^ *export {(.|\n)*${loader}(.|\n)*}`, 'gm'),
        (v: string) => {
          return v
            .replace(new RegExp(`(\\w+ +as +)?${loader}\\W`, 'gm'), (v) =>
              v.endsWith(loader) ? '' : v[v.length - 1]
            )
            .replace(/,( |\n)*,/gm, ',')
            .replace(/{( |\n)*,/gm, '{')
            .replace(/{,( \n)*}/gm, '}')
            .replace(/^ *export +{( |\n)*}\W*$/gm, '')
        }
      )
      // Replacing:
      //    import { something, getStaticProps, somethingelse } from './getStaticProps'
      // To:
      //    import { something, getStaticProps as _getStaticProps, somethingelse } from './getStaticProps'
      .replace(/^ *import +{( |\n)*[^}]*/gm, (v: string) => {
        if (v.match(new RegExp(`\\W+${loader} +as `))) return v
        return v.replace(new RegExp(`\\W+${loader}(\\W|$)`), (r) =>
          r.replace(loader, `${loader} as _${loader}`)
        )
      })
  }

  let template = `
    import __i18nConfig from '@next-translate-root/i18n'
    import __loadNamespaces from 'next-translate/loadNamespaces'
    ${tokenToReplace}
    export async function ${loader}(ctx) {
        ${hasLoader ? `let res = _${loader}(ctx)` : ''}
        ${hasLoader ? `if(typeof res.then === 'function') res = await res` : ''}
        return {
          ${hasLoader ? '...res,' : ''}
          props: {
            ${hasLoader ? '...(res.props || {}),' : ''}
            ...(await __loadNamespaces({
              ...ctx,
              pathname: '${page}',
              loaderName: '${loader}',
              ...__i18nConfig,
              ${overwriteLoadLocales(hasLoadLocaleFrom)}
            }))
          }
        }
    }
  `

  if (typescript) template = template.replace(/\n/g, '\n// @ts-ignore\n')

  // Use callback to avoid parsing special patterns specific for .replace()
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#specifying_a_string_as_a_parameter
  return template.replace(tokenToReplace, () => {
    return `\n${modifiedCode}\n`
  })
}
