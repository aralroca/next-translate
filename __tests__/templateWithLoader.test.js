import templateWithLoader from '../src/plugin/templateWithLoader'
import prettier from 'prettier'

function clean(code) {
  return prettier.format(code, { parser: 'typescript' })
}

const tests = [
  {
    describe: 'exporting as a function - with loader',
    code: `
      export default function Page() {
        return <div>Hello world</div>
      }

      export function getStaticPaths() {
        return {}
      }

      export function getStaticProps(){
        return { props: {} }
      }
  `,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: true,
      },
    ],
  },
  {
    describe: 'exporting as a function - without loader',
    code: `
      export default function Page() {
        return <div>Hello world</div>
      }
  `,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: false,
      },
      {
        page: '/blog/[post]/[...catchall]',
        loader: 'getServerSideProps',
        hasLoader: false,
      },
    ],
  },
  {
    describe: 'exporting as a arrow function - without loader',
    code: `export default () => <div>Hello world</div>`,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: false,
      },
      {
        page: '/blog/[post]/[...catchall]',
        loader: 'getServerSideProps',
        hasLoader: false,
      },
    ],
  },
  {
    describe: 'exporting as a function in diferent line - with loader',
    code: `
      function Page() {
        return <div>Hello world</div>
      }

      function getServerSideProps() {
        return { props: {} }
      }

      export { getServerSideProps }
      export default Page
  `,
    cases: [
      {
        page: '/index',
        loader: 'getServerSideProps',
        hasLoader: true,
      },
    ],
  },
  {
    describe: 'exporting as a function in diferent line - without loader',
    code: `
      function Page() {
        return <div>Hello world</div>
      }

      export default Page
  `,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: false,
      },
      {
        page: '/blog/[post]/[...catchall]',
        loader: 'getServerSideProps',
        hasLoader: false,
      },
    ],
  },
  {
    describe: 'exporting as a class - with loader',
    code: `
      import React from 'react'

      export default class Page extends React.Component {
        render() {
          return <div>Hello world</div>
        }
      }

      export function getStaticProps(){
        return { props: {} }
      }
  `,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: true,
      },
    ],
  },
  {
    describe: 'exporting as a class in a diferent line - without loader',
    code: `
      import React from 'react'

      class Page extends React.Component {
        render() {
          return <div>Hello world</div>
        }
      }

      const myPage = Page

      export default myPage
  `,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: false,
      },
      {
        page: '/blog/[post]/[...catchall]',
        loader: 'getServerSideProps',
        hasLoader: false,
      },
    ],
  },
  {
    describe: 'loader as a arrow function',
    code: `
      export default function Page() {
        return <div>Hello world</div>
      }

      const getStaticProps = () => ({ props: {} })
      export { getStaticProps }
  `,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: true,
      },
    ],
  },
  {
    describe: 'loader imported to another place',
    code: `
      import getStaticProps from 'somewhere/getStaticProps'
      import getStaticPaths from 'somewhere/getStaticPaths'

      const config = {}

      export default function Page() {
        return <div>Hello world</div>
      }

      export { config, getStaticProps, getStaticPaths }
  `,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: true,
      },
    ],
  },
  {
    describe: 'loader with named import to another place',
    code: `
      import { getStaticProps } from 'somewhere/getStaticProps'
      import { getStaticPaths } from 'somewhere/getStaticPaths'

      const config = {}

      export default function Page() {
        return <div>Hello world</div>
      }

      export { config, getStaticProps, getStaticPaths }
  `,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: true,
      },
    ],
  },
  {
    describe: 'loader with one named import to another place',
    code: `
      import { getStaticPaths, getStaticProps, config } from 'somewhere/getStaticProps'

      export default function Page() {
        const test = 'getStaticProps'
        console.log('This should log getStaticProps without any modification')
        return <div>Hello getStaticProps</div>
      }

      export { config, getStaticProps, getStaticPaths }
  `,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: true,
      },
    ],
  },
  {
    describe: 'loader with named import to another place + rename',
    code: `
      import { getStaticPropsA as getStaticProps } from 'somewhere/getStaticProps'
      import { getStaticProps as getStaticPaths } from 'somewhere/getStaticPaths'

      const config = {}

      export default function Page() {
        return <div>Hello world</div>
      }

      export { config, getStaticProps, getStaticPaths }
  `,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: true,
      },
    ],
  },
  {
    describe: 'should add the "as" on the import only when is necessary',
    code: `import { getStaticProps } from "somewhere/getStaticProps";
    import {getStaticProps} from "somewhere/getStaticProps";
    
    import { getStaticPropsA, getStaticProps, getStaticPropsB } from "somewhere/getStaticProps";
    import { getStaticProps as getStaticPaths } from 'somewhere/getStaticPaths'
    import { fake_getStaticProps } from 'somewhere/getStaticPaths'
    
    import { 
      getStaticPropsA, 
      getStaticProps, 
      getStaticPropsB
    } from "somewhere/getStaticProps";
    
    import {
      getStaticPropsA, 
      getStaticProps, 
      getStaticPropsB
    } from "somewhere/getStaticProps";
    let getStaticProps = false
    // Comment to import getStaticProps
    const a = 'import { getStaticProps }'`,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: true,
      },
    ],
  },
  {
    describe: 'should remove exports of existings loaders with "as"',
    code: `export { test as    getStaticProps }
    export {
      test as getStaticProps,
      anotherThing
    }
    
    export {something,getStaticPropsA as getStaticProps}
    export {something,AgetStaticProps as getStaticProps}
    export {something,AgetStaticProps as getStaticProps, getStaticPropsB}`,
    cases: [
      {
        page: '/index',
        loader: 'getStaticProps',
        hasLoader: true,
      },
    ],
  },
]

describe('templateWithLoader', () => {
  tests.forEach((d) => {
    describe(d.describe, () => {
      d.cases.forEach(({ expected, debug, ...options }) => {
        const fn = debug ? test.only : test
        const testname = Object.entries(options).map(([k, v]) => `${k}: ${v}`)
        fn(testname.join(' | '), () => {
          expect(clean(templateWithLoader(d.code, options))).toMatchSnapshot()
        })
      })
    })
  })
})
