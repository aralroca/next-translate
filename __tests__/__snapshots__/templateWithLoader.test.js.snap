// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templateWithLoader exporting as a arrow function - without loader page: /blog/[post]/[...catchall] | loader: getServerSideProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');
export default () => <div>Hello world</div>;

export async function getServerSideProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/blog/[post]/[...catchall]\\",
        loaderName: \\"getServerSideProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a arrow function - without loader page: /index | loader: getStaticProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');
export default () => <div>Hello world</div>;

export async function getStaticProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a class - with loader page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

export default class Page extends React.Component {
  render() {
    return <div>Hello world</div>;
  }
}

function _getStaticProps() {
  return { props: {} };
}

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a class in a diferent line - without loader page: /blog/[post]/[...catchall] | loader: getServerSideProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

class Page extends React.Component {
  render() {
    return <div>Hello world</div>;
  }
}

const myPage = Page;

export default myPage;

export async function getServerSideProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/blog/[post]/[...catchall]\\",
        loaderName: \\"getServerSideProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a class in a diferent line - without loader page: /index | loader: getStaticProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

class Page extends React.Component {
  render() {
    return <div>Hello world</div>;
  }
}

const myPage = Page;

export default myPage;

export async function getStaticProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function - with loader page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

export default function Page() {
  return <div>Hello world</div>;
}

export function getStaticPaths() {
  return {};
}

function _getStaticProps() {
  return { props: {} };
}

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function - without loader page: /blog/[post]/[...catchall] | loader: getServerSideProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

export default function Page() {
  return <div>Hello world</div>;
}

export async function getServerSideProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/blog/[post]/[...catchall]\\",
        loaderName: \\"getServerSideProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function - without loader page: /index | loader: getStaticProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

export default function Page() {
  return <div>Hello world</div>;
}

export async function getStaticProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function in diferent line - with loader page: /index | loader: getServerSideProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

function Page() {
  return <div>Hello world</div>;
}

function _getServerSideProps() {
  return { props: {} };
}

export default Page;

export async function getServerSideProps(ctx) {
  let res = _getServerSideProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getServerSideProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function in diferent line - without loader page: /blog/[post]/[...catchall] | loader: getServerSideProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

function Page() {
  return <div>Hello world</div>;
}

export default Page;

export async function getServerSideProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/blog/[post]/[...catchall]\\",
        loaderName: \\"getServerSideProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function in diferent line - without loader page: /index | loader: getStaticProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

function Page() {
  return <div>Hello world</div>;
}

export default Page;

export async function getStaticProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader loader as a arrow function page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

export default function Page() {
  return <div>Hello world</div>;
}

const _getStaticProps = () => ({ props: {} });

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader loader imported to another place page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import _getStaticProps from \\"somewhere/getStaticProps\\";
import getStaticPaths from \\"somewhere/getStaticPaths\\";

const config = {};

export default function Page() {
  return <div>Hello world</div>;
}

export { config, getStaticPaths };

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader loader with named import to another place + rename page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import { getStaticPropsA as _getStaticProps } from \\"somewhere/getStaticProps\\";
import { getStaticProps as getStaticPaths } from \\"somewhere/getStaticPaths\\";

const config = {};

export default function Page() {
  return <div>Hello world</div>;
}

export { config, getStaticPaths };

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader loader with named import to another place page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import { getStaticProps as _getStaticProps } from \\"somewhere/getStaticProps\\";
import { getStaticPaths } from \\"somewhere/getStaticPaths\\";

const config = {};

export default function Page() {
  return <div>Hello world</div>;
}

export { config, getStaticPaths };

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader loader with one named import to another place page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import {
  getStaticPaths,
  getStaticProps as _getStaticProps,
  config,
} from \\"somewhere/getStaticProps\\";

export default function Page() {
  const test = \\"getStaticProps\\";
  console.log(\\"This should log getStaticProps without any modification\\");
  return <div>Hello getStaticProps</div>;
}

export { config, getStaticPaths };

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader should add the "as" on the import only when is necessary page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');
import { getStaticProps as _getStaticProps } from \\"somewhere/getStaticProps\\";
import { getStaticProps as _getStaticProps } from \\"somewhere/getStaticProps\\";

import {
  getStaticPropsA,
  getStaticProps as _getStaticProps,
  getStaticPropsB,
} from \\"somewhere/getStaticProps\\";
import { getStaticProps as getStaticPaths } from \\"somewhere/getStaticPaths\\";
import { fake_getStaticProps } from \\"somewhere/getStaticPaths\\";

import {
  getStaticPropsA,
  getStaticProps as _getStaticProps,
  getStaticPropsB,
} from \\"somewhere/getStaticProps\\";

import {
  getStaticPropsA,
  getStaticProps as _getStaticProps,
  getStaticPropsB,
} from \\"somewhere/getStaticProps\\";
let _getStaticProps = false;
// Comment to import getStaticProps
const a = \\"import { getStaticProps }\\";

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader should remove exports of existings loaders with "as" page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

export { anotherThing };

export { something };
export { something };
export { something, getStaticPropsB };

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        loaderName: \\"getStaticProps\\",
        ...__i18nConfig,
        loadLocaleFrom: (l, n) =>
          import(\`@next-translate-root/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;
